### ISSUE TITLE
Without verification of whether the B type is legal may lead to data inconsistency in alternator

### ISSUE THREAD

**Issue by Felix-zhoux** [scylladb/scylladb#17539]
For binary types, **dynamodb** requires "Your applications must encode binary values in base64-encoded format before sending them to DynamoDB." (ref: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes. html#HowItWorks.DataTypes)

Using alternator, if we use a `B` type that does not conform to the **base64 encoding format**, the successfully written data may not be as expected.

In the following example, you can find that the `"c": {"B": "fakebase64"}}` planned to be written actually becomes `"c":{"B":"fakebase6w=="}`

```
// send a updateitem request to scylla
$ curl -v \
>   http://127.0.0.1:8000 \
>   -H 'Content-Type: application/json' \
>   -H 'X-Amz-Target: DynamoDB_20120810.UpdateItem' \
>   -H "Authorization: AWS4-HMAC-SHA256 Credential=mykey/20220226/us-east-1/dynamodb/aws4_request, SignedHeaders=host;x-amz-date, Signature=39c31b71f191f8610c5fa429d49b25273db20e7512ce7f4adf495bda67fe2ac4"\
>   -d '{"ReturnValues": "ALL_OLD", "TableName": "test_aws", "UpdateExpression": "SET c = :v", "Key": {"sk": {"S": "sk1"}, "pk": {"S": "pk1"}}, "ExpressionAttributeValues": {":v": {"B": "fakebase64"}}}'
* About to connect() to 127.0.0.1 port 8000 (#0)
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> POST / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 127.0.0.1:8000
> Accept: */*
> Content-Type: application/json
> X-Amz-Target: DynamoDB_20120810.UpdateItem
> Authorization: AWS4-HMAC-SHA256 Credential=mykey/20220226/us-east-1/dynamodb/aws4_request, SignedHeaders=host;x-amz-date, Signature=39c31b71f191f8610c5fa429d49b25273db20e7512ce7f4adf495bda67fe2ac4
> Content-Length: 193
> 
* upload completely sent off: 193 out of 193 bytes
< HTTP/1.1 200 OK
< Content-Length: 2
< Content-Type: application/x-amz-json-1.0
< Date: Tue, 27 Feb 2024 06:19:14 GMT
< Server: Seastar httpd
< 
* Connection #0 to host 127.0.0.1 left intact
{}

// getitem from scylla
$ curl -v \
>   http://127.0.0.1:8000 \
>   -H 'Content-Type: application/json' \
>   -H 'X-Amz-Target: DynamoDB_20120810.GetItem' \
>   -H "Authorization: AWS4-HMAC-SHA256 Credential=mykey/20220226/us-east-1/dynamodb/aws4_request, SignedHeaders=host;x-amz-date, Signature=39c31b71f191f8610c5fa429d49b25273db20e7512ce7f4adf495bda67fe2ac4"\
>   -d '{"TableName": "test_aws", "Key": {"sk": {"S": "sk1"}, "pk": {"S": "pk1"}}}'
* About to connect() to 127.0.0.1 port 8000 (#0)
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> POST / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 127.0.0.1:8000
> Accept: */*
> Content-Type: application/json
> X-Amz-Target: DynamoDB_20120810.GetItem
> Authorization: AWS4-HMAC-SHA256 Credential=mykey/20220226/us-east-1/dynamodb/aws4_request, SignedHeaders=host;x-amz-date, Signature=39c31b71f191f8610c5fa429d49b25273db20e7512ce7f4adf495bda67fe2ac4
> Content-Length: 74
> 
* upload completely sent off: 74 out of 74 bytes
< HTTP/1.1 200 OK
< Content-Length: 69
< Content-Type: application/x-amz-json-1.0
< Date: Tue, 27 Feb 2024 06:20:06 GMT
< Server: Seastar httpd
< 
* Connection #0 to host 127.0.0.1 left intact
{"Item":{"pk":{"S":"pk1"},"sk":{"S":"sk1"},"c":{"B":"fakebase6w=="}}}
```

However, if you use **dynamodb** (tested using dynamodb_local), although the error message is not clear (`HTTP/1.1 500 Server Error, "message":"The request processing has failed because of an unknown error, exception or failure."`), data that does not conform to the base64 encoding format **will not be written** successfully, and the request will fail.

When the same request is sent to dynamodb, an error will be reported.
```
# curl -v \
>   http://127.0.0.1:8080 \
>   -H 'Content-Type: application/json' \
>   -H 'X-Amz-Target: DynamoDB_20120810.UpdateItem' \
>   -H "Authorization: AWS4-HMAC-SHA256 Credential=mykey/20220226/us-east-1/dynamodb/aws4_request, SignedHeaders=host;x-amz-date, Signature=39c31b71f191f8610c5fa429d49b25273db20e7512ce7f4adf495bda67fe2ac4"\
>   -d '{"ReturnValues": "ALL_OLD", "TableName": "test_aws", "UpdateExpression": "SET c = :v", "Key": {"sk": {"S": "sk1"}, "pk": {"S": "pk1"}}, "ExpressionAttributeValues": {":v": {"B": "fakebase64"}}}'
* About to connect() to 127.0.0.1 port 8080 (#0)
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> POST / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 127.0.0.1:8080
> Accept: */*
> Content-Type: application/json
> X-Amz-Target: DynamoDB_20120810.UpdateItem
> Authorization: AWS4-HMAC-SHA256 Credential=mykey/20220226/us-east-1/dynamodb/aws4_request, SignedHeaders=host;x-amz-date, Signature=39c31b71f191f8610c5fa429d49b25273db20e7512ce7f4adf495bda67fe2ac4
> Content-Length: 193
> 
* upload completely sent off: 193 out of 193 bytes
< HTTP/1.1 500 Server Error
< Content-Type: application/x-amz-json-1.0
< x-amzn-RequestId: 62e6fd48-99ee-4db7-9011-3ff5bc64db2f
< Content-Length: 158
< Server: Jetty(8.1.12.v20130726)
< 
* Connection #0 to host 127.0.0.1 left intact
{"__type":"com.amazonaws.dynamodb.v20120810#InternalFailure","message":"The request processing has failed because of an unknown error, exception or failure."
```

So, it looks like we need to do some format checking for the binary type in the alternator?

**Comment by nyh** [comment]
We have in test/alternator/test_manual_request.py some tests with invalid base64 and seeing them fail, but I see now that these tests were with PutItem and FilterExpression, and perhaps the code path in UpdateItem with UpdateExpression - which you used - is different and missing the error check?

**Comment by nyh** [comment]
Several different requests (such as PutItem, BatchPutItem, UpdateItem with AttributeUpdates or UpdateExpression) can take be used to write a value, encoded in DynamoDB's JSON format such as `{"S": "hello"}`. While we could blindly copy this value into storage and then blindly read it back when the item is read, this would be a mistake - we need to validate it's legally encoded value. We have a function to do exactly this - `validate_value()`, and we call it in all of the above cases except the UpdateExpression case where we forgot.

The most difficult part of fixing this is writing the test, which isn't easy (boto3 doesn't help us to write tests with malformed parameters).

**Comment by nyh** [comment]
I mispoke: Actually, the UpdateExpression also had validation in place - the function resolve_constant() which resolves a reference to the JSON value, does call validate_value() as it should.
But there is some bug somewhere, the error we get is "Invalid escape character in string. at 56211208" (?). I'm debugging it. I already have working tests.

**Comment by nyh** [comment]
@Felix-zhoux please check this issue again. I think it is actually fixed already and should be closed. I'll prepare a test that validates it is fixed.

**Comment by nyh** [comment]
I verified with my new test (it's so easy with `test/alternator/run --release`!) that this bug existed in Scylla 5.2 but fixed in Scylla 5.3. Long ago. I guess you're using an old version of Scylla? 
I'm closing this issue. Please reopen if you think I made a mistake.

**Comment by nyh** [comment]
By the way a comment for posterity:
This issue was probably a duplicate of https://github.com/scylladb/scylladb/issues/6487 (that issue focused on missing padding, but other problems in encoding are possible), fixed by @nuivall in https://github.com/scylladb/scylladb/pull/11944. 
The fix commit https://github.com/scylladb/scylladb/commit/48e2d6a54113b0b17266856c912a1933cdbc5d72 reached the  5.3 branch - and was never backported to earlier branches.
This explains my conclusions above about the branches vulnerable to this bug.
